### PR TITLE
Assigning Unicode string objects to string datasets

### PR DESCRIPTION
**Pull request by ajelenak-thg**
These changes fix this issue: HDFGroup/hpd-ws#26.

I traced the problem to assignment of a Unicode string object, e.g. ``u"Hello world!"`` to an already created dataset of the ASCII datatype (per its JSON information). When the json package converts JSON to Python dictionary all strings, its keys and string values, are Unicode string objects.

The crux of the changes are:

1. Allow conversion of every string value from JSON into the Python's str object.
1. Do not accept H5T_STRING object values from JSON as-is.

I would keep this branch alive for testing when assigning str or unicode objects to UTF-8 string datasets.

### CODE REVIEW

**Comment by jreadey** [comment]
It's hard to sort out the relevant change among all the white space updates.  Is just around line 1066 that has the relevant update?

File: server/hdf5db.py [hunk 1]
```diff
@@ -1062,43 +1062,45 @@ def getDataValue(self, typeItem, value, dimension=0, dims=None):
             rank = len(type_dims)
             baseType = typeItem['base']
             out = self.getDataValue(baseType, value, dimension=rank, dims=type_dims)
-            
-        elif typeClass in ('H5T_INTEGER', 'H5T_FLOAT', 'H5T_STRING', 'H5T_ENUM'):
+
```
Comment (ajelenak-thg) [inline server/hdf5db.py]: Here!

File: util/jsontoh5.py [hunk 2]
```diff
@@ -86,15 +86,16 @@ def createDataset(self, uuid, body):
...
         self.db.createDataset(datatype, dims, max_shape=max_shape, fill_value=fill_value,
             obj_uuid=uuid)
-            
+
         if "value" in body:
             data = body["value"]
             if data:
-                self.db.setDatasetValuesByUuid(uuid, data) 
-            
+                data = self.db.toRef(len(dims), datatype, data)
```
Comment (ajelenak-thg) [inline util/jsontoh5.py]: Here!

**Comment by ajelenak-thg** [comment]
I apologize for the "white noise" of changes. Can you configure your text editor to remove all the excess white space?

I put "Here!" to mark relevant code changes.

**Comment by ajelenak-thg** [comment]
I just checked h5py's tests and fixed-length Unicode datasets are not supported. Vlen Unicode datasets are.

**Comment by jreadey** [comment]
test script passes ok, went ahead and merged to master.

**Pull request merged**